### PR TITLE
Ports - Stephanie

### PR DESCRIPTION
**Pull request by smarchante1**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? |  One major challenge I had was deciding if I wanted to keep a date range class that I originally had. I ended up removing it and think that might not have been the best decision.
What was a design decision you made that changed over time over the project? | Initially I wanted to store block reservations in my main Reservation Manager class. I changed this to storing block reservations in a hotel block class.
What was a concept you gained clarity on, or a learning that you'd like to share? | I have a better understanding of how classes can interact with one another through the methods that are created in each class. This also made the project quite challenging.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | A nominal test I wrote was making sure a reservation was added to the list of reservations.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | An edge case I wrote was checking to make sure a check out date wasn't before a check_in date.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I really enjoyed this workflow. It helped me to get a better understanding of the methods I needed, and then slowly work towards writing them.

### CODE REVIEW

File: spec/reservation_manager_spec.rb [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    end
+     
+    it "stores booked reservations in reservations array" do
+      res1 = @frontdesk.book_reservation(1, @test_range)
+      @reservations << res1
+
+      expect(@reservations.length).must_equal(1)    
+    end
+
+    it "raises an error if the room is not available" do
```
Comment (CheezItMan) [inline spec/reservation_manager_spec.rb:60]: Your room here is still available so the `ArgumentError` is not being raised.  This is a good start on testing the double-book issue.

File: lib/reservation_manager.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      return rooms 
+    end
+
+    #  basic reservation methods
+    def book_reservation(room, date_range)
+      # raising the error but test says nothing was raised
+      # unless room_availability(date_range.check_in, date_range.check_out).include?(room)
+      #   raise ArgumentError, "The room you've selected is not available for these dates."
+      # end
+      reservation = Hotel::Reservation.new(room, date_range)
```
Comment (CheezItMan) [inline lib/reservation_manager.rb:23]: No check that the room is already reserved.

File: lib/date_range.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
+require 'date'
+
+module Hotel
+  class DateRange
+    attr_reader :check_in, :check_out
+    
+    def initialize(check_in, check_out)
+      @check_in = check_in
+  
+      if check_out < check_in
```
Comment (CheezItMan) [inline lib/date_range.rb:10]: You're checking this twice, here and with the `total_stay` method.  A little odd.  Also this if statement doesn't check to see if the dates are equal, but your total_stay method did.

File: lib/date_range.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      end
+
+    end
+
+    def total_stay
+      total_nights = (check_out - check_in).to_i
+      return total_nights
+    end
+
+    def date_check(date)
```
Comment (CheezItMan) [inline lib/date_range.rb:26]: This method is a little vaguely named.  I suggest naming it `within_range` to tell if a date is within the DateRange.

File: lib/date_range.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      valid_date = (date >= @check_in && date < @check_out)
+      return valid_date
+    end
+    
+    def daterange_check(check_in, check_out)
```
Comment (CheezItMan) [inline lib/date_range.rb]: This method is also a little misleading, I would call it `overlap` and pass another `DateRange` instance to compare.  If this is the method you're using and you don't expect users to use `date_check` directly, I would also make that method private.

File: lib/hotel_block.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
+module Hotel
+  class HotelBlock
+    attr_reader :room, :check_in, :check_out, :cost, :reservations, :hotel
```
Comment (CheezItMan) [inline lib/hotel_block.rb]: Looks like you got started here, but didn't get far.  

Just one observation, I think the `hotel` instance method is unneeded.  You should have something to keep track of the rooms in the block however.

File: lib/reservation.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  
+    def initialize(room, date_range, cost: 200)
+      @room = room
+      @date_range = date_range
+      @cost = cost
+
+    end
+
+    def total_cost
+      if cost == 200
```
Comment (CheezItMan) [inline lib/reservation.rb:14]: Why this if statement here?  Why not just use the value of `cost`?

File: lib/reservation_manager.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
+require 'date'
+
+module Hotel
+  class ReservationManager
+    attr_reader :reservations, :block_reservations
+
+    def initialize
+      @reservations = []
+      @block_reservations = []
+    end
```
Comment (CheezItMan) [inline lib/reservation_manager.rb:11]: I suggest also keeping `rooms` as an array of numbers and setting it with your `hotel_rooms` method, so you don't have to repeatably call that method.

File: spec/date_range_spec.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  describe 'total stay method' do
+    it 'calculates the total stay' do
+      expect(@test_daterange.total_stay).must_equal(4)
+    end
+  end
+
+  describe 'date check method' do
+    it 'returns true if checked date is within daterange (false if not)' do
+      expect(@test_daterange.date_check(Date.new(2019,3,12))).must_equal(true)
```
Comment (CheezItMan) [inline spec/date_range_spec.rb:38]: You should also check the `start_date` and `end_date` and check to see `date_check` on those dates.

File: spec/date_range_spec.rb [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      expect(@test_daterange.date_check(Date.new(2019,3,14))).must_equal(true)
+      expect(@test_daterange.date_check(Date.new(2019,3,20))).must_equal(false)
+      expect(@test_daterange.date_check(Date.new(2019,3,15))).must_equal(false)
+    end
+  end
+
+  describe 'daterange check method' do
+    it 'returns true if checked date is within daterange (false if not)' do
+      expect(@test_daterange.daterange_check(Date.new(2019,3,12),Date.new(2019,3,14))).must_equal(true)
```
Comment (CheezItMan) [inline spec/date_range_spec.rb]: Similar to the above, check the edge cases, date ranges that end on the start date of your date_range and another that starts on the end_date.  

Then try some overlaps.

File: spec/reservation_manager_spec.rb [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+        @frontdesk.book_reservation(room, Hotel::DateRange.new(Date.new(2019,3,15),Date.new(2019,3,18)))
+      end
+      expect(@frontdesk.room_availability(Date.new(2019,3,15), Date.new(2019,3,18))).must_equal([*6..20])
+    end
+
+    it "returns empty array if all rooms are booked" do
+      [*1..20].each do |room|
+        @frontdesk.book_reservation(room, Hotel::DateRange.new(Date.new(2019,3,15),Date.new(2019,3,18)))
+      end
+      expect(@frontdesk.room_availability(Date.new(2019,3,15), Date.new(2019,3,18))).must_equal([])
```
Comment (CheezItMan) [inline spec/reservation_manager_spec.rb:33]: 👍

File: spec/reservation_manager_spec.rb [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      @frontdesk.book_reservation(5,
+        Hotel::DateRange.new(Date.new(2019,3,5), Date.new(2019,3,8)))
+      @res_list = @frontdesk.res_by_date(Date.new(2019,3,7))
+    end
+
+    it 'returns an empty array if there are no reservations during selected date' do
+      expect(@frontdesk.res_by_date(Date.new(2002,3,15)) ).must_be_instance_of(Array)
+    end
+
+    it "returns array of all booked reservations" do
```
Comment (CheezItMan) [inline spec/reservation_manager_spec.rb:96]: You should also test that it has the given reservations.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Good number of commits and good commit messages
Answer comprehension questions | Check 
**Design** |
Each class is responsible for a single piece of the program | For what's been done yes.
Classes are loosely coupled | Check
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check, but see my inline notes, the if statement is... odd
Invalid date range produces an error | Check
Test coverage | 88.44% coverage :(
**Wave 2** |
View available rooms for a given date range | **Missing**
Reserving a room that is not available produces an error | **Missing**
**Wave 3** |
Create a block of rooms | **Missing**
Check if a block has rooms | **Missing**
Reserve a room from a block | **Missing**
**Fundamentals** |
Names variables, classes and modules appropriately | See my inline notes for some of your method names.  
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Check
Appropriately uses methods to break down tasks into smaller simpler tasks | Check, for what was done
Understands the differences between class and instance methods | Only instance methods here. 
Appropriately uses iterators and `Enumerable` methods | Very few examples of Enumerable methods here. 
Appropriately writes and utilizes classes | Check
Appropriately utilizes modules as a namespace | Check
**Wrap Up** |
There is a refactors.txt file | Check
The file provides a roadmap to future changes | It's starting to
**Additional Feedback** | As we noted in a 1-on-1, this was a rough project for you.  Not a great outcome, but you had the bones of a great project here.  You had good classes and a good rough outline of your design.  Unfortunately it wasn't implemented and that's unfortunate.  Take a look at my feedback and ask me questions if something is unclear.  This is a tough industry, but you do have the talent to be very successful here.  Don't let what other people say distract you.  You can make a great impact here!  If you are feeling down and upset please talk to me, Sarah or Dee and let us support you!

**Comment by smarchante1** [comment]
Hi Chris!

I think I might have misunderstood the Hotel Revisted project. I submitted
my project when it was due and thought that we were supposed to make
changes even though we didn’t receive feedback, to start the break week
refactor project. I commented out the original submission for my project
and I was working on the refactor with date ranges today under the
impression that it was break week homework. If you look at the original
commit, that was my original submission going into wave 3. Is it possible
for you to take a look at it?
